### PR TITLE
CSHARP-4420: EG tests for Atlas Search

### PR DESCRIPTION
**Pull request by BorisDog**
This PR is rebased on top of [CSHARP-4440](https://github.com/mongodb/mongo-csharp-driver/pull/989).
The main focus here is[ AtlasSearchTest.cs](https://github.com/mongodb/mongo-csharp-driver/blob/cb8a4be121c036a315f31eb7c43dda712e12116a/tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs)
Please review just the last [commit](https://github.com/mongodb/mongo-csharp-driver/pull/1001/commits/cb8a4be121c036a315f31eb7c43dda712e12116a).

### CODE REVIEW

File: evergreen/run-atlas-search-test.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
+
+set -o xtrace
+set -o errexit  # Exit the script with error if any of the commands fail
```
Comment (DmitryLukyanov) [inline evergreen/run-atlas-search-test.sh:4]: add a section like in most of other files about output variables: https://github.com/mongodb/mongo-csharp-driver/blob/master/evergreen/run-csfle-azure-tests.sh#L11

File: src/MongoDB.Driver/Search/Highlight.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    /// <summary>
+    /// Represents a result of highlighting.
+    /// </summary>
+    public sealed class Highlight
+    {
+        /// <summary>
+        /// Gets or sets the document field which returned a match.
+        /// </summary>
+        [BsonElement("path")]
+        public string Path { get; private set; }
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/Search/Highlight.cs:29]: why `private set` is here? Sounds like it does nothing?

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+        {
+            RequireEnvironment.Check().EnvironmentVariable("ATLAS_SEARCH_TESTS_ENABLED");
+
+            var atlasSearchUri = Environment.GetEnvironmentVariable("ATLAS_SEARCH");
+            Ensure.IsNotNullOrEmpty(atlasSearchUri, nameof(atlasSearchUri));
+
+            _disposableMongoClient = new(new MongoClient(atlasSearchUri), CreateLogger<DisposableMongoClient>());
+        }
+
+        protected override void DisposeInternal() =>
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs:65]: nit: I think all content suites in one line like: 
` protected override void DisposeInternal() => _disposableMongoClient.Dispose();`

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+            var result = SearchSingle(Builders.Search.Wildcard("tranquil*", x => x.Body, true));
+
+            result.Title.Should().Be("US Constitution");
+        }
+
+        private List<AirbnbListing> GeoSearch(SearchDefinition<AirbnbListing> searchDefintion) =>
+            GetGeoTestCollection().Aggregate().Search(searchDefintion).ToList();
+
+        private HistoricalDocument SearchSingle(SearchDefinition<HistoricalDocument> searchDefintion) =>
+            GetTestCollection().Aggregate().Search(searchDefintion)
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs:428]: this should be at least:

     GetTestCollection().Aggregate().Search(searchDefintion)
        .Limit(1)
        .ToList()
        .Single();

but I would recommend a better formatting:

     GetTestCollection()
        .Aggregate()
        .Search(searchDefintion)
        .Limit(1)
        .ToList()
        .Single();

File: src/MongoDB.Driver/Search/Highlight.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Search/Highlight.cs:29]: sounds like it's for deserialization only. Does it deserve a comment?

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+using Xunit.Abstractions;
+using Builders = MongoDB.Driver.Builders<MongoDB.Driver.Tests.Search.AtlasSearchTest.HistoricalDocument>;
+using GeoBuilders = MongoDB.Driver.Builders<MongoDB.Driver.Tests.Search.AtlasSearchTest.AirbnbListing>;
+
+namespace MongoDB.Driver.Tests.Search
+{
+    [Trait("Category", "AtlasSearch")]
+    public class AtlasSearchTest : LoggableTestClass
+    {
+        private static readonly GeoJsonPolygon<GeoJson2DGeographicCoordinates> __testPolygon =
```
[the remark this answers is not in the document]
Comment (BorisDog) [reply tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs:40]: Done.

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs [hunk 6]
```diff
@@ -0,0 +1,504 @@
...
+            new(new(new(new GeoJson2DGeographicCoordinates[]
```
[the remark this answers is not in the document]
Comment (BorisDog) [reply tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs:41]: This is a long chain of classes creation with long names shortened to omit the names for readability.
So for example in creation of `GeoWithinBox`, the only important and visible data is its 2 corners.

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs (hunk 3 above)
Comment (BorisDog) [reply tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs]: Done.

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs [hunk 7]
```diff
@@ -0,0 +1,504 @@
...
+            Ensure.IsNotNullOrEmpty(atlasSearchUri, nameof(atlasSearchUri));
+
+            _disposableMongoClient = new(new MongoClient(atlasSearchUri), CreateLogger<DisposableMongoClient>());
+        }
+
+        protected override void DisposeInternal() =>
+            _disposableMongoClient.Dispose();
+
+        [Fact]
+        public void Autocomplete()
```
[the remark this answers is not in the document]
Comment (BorisDog) [reply tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs:73]: I am not big fun of having same prefix for all test methods, especially when it does not add any information. Can you think of other alternatives here?

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs [hunk 8]
```diff
@@ -0,0 +1,504 @@
...
+                Builders.Search.Phrase("life, liberty", x => x.Body),
+                Builders.Search.Wildcard("happ*", x => x.Body, true)));
+
+            result.Title.Should().Be("Declaration of Independence");
+        }
+
+        [Fact]
+        public void MustNot()
+        {
+            var result = SearchSingle(Builders.Search.Compound()
```
[the remark this answers is not in the document]
Comment (BorisDog) [reply tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs]: Done.

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs (hunk 4 above)
Comment (BorisDog) [reply tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs]: Done.

File: evergreen/run-atlas-search-test.sh [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+############################################
+#            Main Program                  #
+############################################
+
+echo "Running Atlas Search driver tests"
+
+export ATLAS_SEARCH_TESTS_ENABLED=true
+
+powershell.exe .\\build.ps1 --target TestAtlasSearch
```
[the remark this answers is not in the document]
Comment (BorisDog) [reply evergreen/run-atlas-search-test.sh:17]: Choice for Windows platform is arbitrary, but we don't plan to extend to other OS.

File: evergreen/run-atlas-search-test.sh (hunk 1 above)
Comment (BorisDog) [reply evergreen/run-atlas-search-test.sh:4]: Done.

File: src/MongoDB.Driver/Search/Highlight.cs [hunk 10]
```diff
@@ -0,0 +1,79 @@
+﻿// Copyright 2010-present MongoDB Inc.
```
[the remark this answers is not in the document]
Comment (BorisDog) [reply src/MongoDB.Driver/Search/Highlight.cs]: Good catch, I'll fix this in the main search PR too.

File: src/MongoDB.Driver/Search/Highlight.cs (hunk 2 above)
Comment (BorisDog) [reply src/MongoDB.Driver/Search/Highlight.cs:30]: Yeah it's for deserialization. Looks pretty minor for a comment.

File: src/MongoDB.Driver/Search/Highlight.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Search/Highlight.cs]: for the future :( I would suggest using a different PR approach (we already did similar work in past) since it's hard to follow what was fixed and what should be reviewed. Sounds like at least in this file the above suggestion was applied..

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs [hunk 11]
```diff
@@ -0,0 +1,504 @@
...
+        private readonly DisposableMongoClient _disposableMongoClient;
+
+        public AtlasSearchTest(ITestOutputHelper testOutputHelper) : base(testOutputHelper)
+        {
+            RequireEnvironment.Check().EnvironmentVariable("ATLAS_SEARCH_TESTS_ENABLED");
+
+            var atlasSearchUri = Environment.GetEnvironmentVariable("ATLAS_SEARCH");
+            Ensure.IsNotNullOrEmpty(atlasSearchUri, nameof(atlasSearchUri));
+
+            _disposableMongoClient = new(new MongoClient(atlasSearchUri), CreateLogger<DisposableMongoClient>());
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs:67]: I would suggest creating a new CreateDisposableClient(connectionString) method to DriverTestConfiguration to be able creating a test client in the same way across driver tests

**Review by DmitryLukyanov — COMMENTED** [review]
some mostly minor comments

File: evergreen/run-atlas-search-test.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (BorisDog) [reply evergreen/run-atlas-search-test.sh:17]: I think it's fine for now, just copied from `run-atlas-data-lake-test` and `run-atlas-connectivity-tests`

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs (hunk 11 above)
Comment (BorisDog) [reply tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs:67]: I would suggest doing this in different small refactoring PR, to change all other places too.

**Review by DmitryLukyanov — APPROVED** [review]
LGTM

File: src/MongoDB.Driver/Search/Highlight.cs [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+using MongoDB.Bson;
+using MongoDB.Bson.Serialization.Attributes;
+
+namespace MongoDB.Driver.Search
+{
+    /// <summary>
+    /// Represents a result of highlighting.
+    /// </summary>
+    public sealed class Highlight
```
Comment (rstam) [inline src/MongoDB.Driver/Search/Highlight.cs:24]: Why is this new file in this PR? Shouldn't all work that is not testing related be in the CSHARP-4440 PR?

In any case, I'm not reviewing this file in this PR because it's not testing related.

This whole folder is wildly out of date with CSHRP-4440.

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs [hunk 13]
```diff
@@ -0,0 +1,519 @@
...
+using MongoDB.TestHelpers.XunitExtensions;
+using Xunit;
+using Xunit.Abstractions;
+using Builders = MongoDB.Driver.Builders<MongoDB.Driver.Tests.Search.AtlasSearchTest.HistoricalDocument>;
+using GeoBuilders = MongoDB.Driver.Builders<MongoDB.Driver.Tests.Search.AtlasSearchTest.AirbnbListing>;
+
+namespace MongoDB.Driver.Tests.Search
+{
+    [Trait("Category", "AtlasSearch")]
+    public class AtlasSearchTest : LoggableTestClass
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs:36]: nit: rename file and class to `AtlasSearchTests` (plural)

File: tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (rstam) [reply tests/MongoDB.Driver.Tests/Search/AtlasSearchTest.cs:41]: > omit the names for readability

LOL. 

I found it much *harder* to read because there is no way to know what each call to `new` is doing without going into Visual Studio and hovering the mouse over each `new` to see what it does.

I agree with Dima. This looks like an abuse of a new C# feature.

**Review by rstam — CHANGES_REQUESTED** [review]
Minor changes requested. Overall looks good.

I think this needs to be reviewed again (in a new clean PR) once you have rebased this on the csharp4440 work.

File: evergreen/run-atlas-search-test.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (DmitryLukyanov) [reply evergreen/run-atlas-search-test.sh:17]: it's mandatory only on non windows OSs if I recall correctly

[the remark this answers is not in the document]
Comment (rstam) [reply evergreen/run-atlas-search-test.sh:17]: At one point it was mandatory on my Windows machine also, so I got used to using the `=` form.

My guess is that at some point a Cake release made it mandatory, and they got lots of complaints, and then they reverted the change.

I'm assuming it's Cake that parses this argument(s) and not Powershell.
